drivers/dma/dma_nios2_msgdma.c

```diff
@@ -0,0 +1,240 @@
+	}
+
+#if MSGDMA_0_CSR_PREFETCHER_ENABLE
+	if (cfg->block_count > 1) {
+		SYS_LOG_ERR("driver yet add support multiple descriptors");
+		return -EINVAL;
+	}
+#else
+	if (cfg->block_count != 1) {
+		SYS_LOG_ERR("invalid block count!!");
+		return -EINVAL;
+	}
+#endif
+
+	if (cfg->head_block == NULL) {
+		SYS_LOG_ERR("head_block ptr NULL!!");
+		return -EINVAL;
+	}
+
+	if (cfg->head_block->block_size > MSGDMA_0_DESCRIPTOR_SLAVE_MAX_BYTE) {
+		SYS_LOG_ERR("DMA error: Data size too big: %d",
+			    cfg->head_block->block_size);
+		return -EINVAL;
+	}
+
+	k_sem_take(&dev_cfg->sem_lock, K_FOREVER);
+	dev_cfg->dma_callback = cfg->dma_callback;
+	dev_cfg->direction = cfg->channel_direction;
+	dma_block = cfg->head_block;
+	control =  ALTERA_MSGDMA_DESCRIPTOR_CONTROL_TRANSFER_COMPLETE_IRQ_MASK |
+		  ALTERA_MSGDMA_DESCRIPTOR_CONTROL_EARLY_TERMINATION_IRQ_MASK;
+
+	/*
+	 * Suppress -Wint-conversion compiler warnings which
+	 * are caused due to the type mismatch b/w the dma_block
+	 * source/destination address (u32_t) and HAL driver's
+	 * descriptor construction API's arguments (alt_32 *)
+	 */
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wint-conversion"
+	if (dev_cfg->direction == MEMORY_TO_MEMORY) {
+		status = alt_msgdma_construct_standard_mm_to_mm_descriptor(
+			dev_cfg->msgdma_dev, &dev_cfg->desc,
+			dma_block->source_address,
+			dma_block->dest_address,
+			dma_block->block_size,
+			control);
+	} else if (dev_cfg->direction == MEMORY_TO_PERIPHERAL) {
+		status = alt_msgdma_construct_standard_mm_to_st_descriptor(
+			dev_cfg->msgdma_dev, &dev_cfg->desc,
+			dma_block->source_address,
+			dma_block->block_size,
+			control);
+	} else if (dev_cfg->direction == PERIPHERAL_TO_MEMORY) {
+		status = alt_msgdma_construct_standard_st_to_mm_descriptor(
+			dev_cfg->msgdma_dev, &dev_cfg->desc,
+			dma_block->dest_address,
+			dma_block->block_size,
+			control);
+	} else {
+		SYS_LOG_ERR("invalid channel direction");
+		status = -EINVAL;
+	}
+#pragma GCC diagnostic pop
+
+	/* Register msgdma callback */
+	alt_msgdma_register_callback(dev_cfg->msgdma_dev,
+			nios2_msgdma_callback,
+			ALTERA_MSGDMA_CSR_GLOBAL_INTERRUPT_MASK |
+			ALTERA_MSGDMA_CSR_STOP_ON_ERROR_MASK |
+			ALTERA_MSGDMA_CSR_STOP_ON_EARLY_TERMINATION_MASK,
+			dev);
+
+	/* Clear the IRQ status */
+	IOWR_ALTERA_MSGDMA_CSR_STATUS(dev_cfg->msgdma_dev->csr_base,
+				      ALTERA_MSGDMA_CSR_IRQ_SET_MASK);
+	k_sem_give(&dev_cfg->sem_lock);
+
+	return status;
+}
+
+static int nios2_msgdma_transfer_start(struct device *dev, u32_t channel)
+{
+	struct nios2_msgdma_dev_cfg *cfg = DEV_CFG(dev);
+	int status;
+
+	/* Nios-II mSGDMA supports only one channel per DMA core */
+	if (channel != 0) {
+		SYS_LOG_ERR("Invalid channel number");
+		return -EINVAL;
+	}
+
+	k_sem_take(&cfg->sem_lock, K_FOREVER);
+	status = alt_msgdma_standard_descriptor_async_transfer(cfg->msgdma_dev,
+								&cfg->desc);
+	if (status < 0) {
+		SYS_LOG_ERR("DMA transfer error (%d)", status);
```

Comment:
could move this outside the semaphore take/give section

Reply:
Ok